server/app/models.py

```diff
@@ -658,15 +660,50 @@ def mark_as_final(self):
             final = FinalSubmission.query(
                 FinalSubmission.assignment==assignment,
                 FinalSubmission.submitter==submitter).get()
+        return final
+
+    def mark_as_final(self):
+        """Create or update a final submission."""
+        final = self.get_final()
+
         if final:
             final.submitter = self.submitter
             final.submission = self.key
         else:
-            final = FinalSubmission(assignment=assignment, submission=self.key)
+            group = self.submitter.get().get_group(self.assignment)
+            final = FinalSubmission(
+                assignment=self.assignment, submission=self.key)
             if group:
                 final.group = group.key
         final.put()
 
+    def resubmit(self, user_key):
```

Comment:
Hmm... I'm torn. On one hand, I think there's a way to write this so that we don't duplicate code. On the other hand, I'm _really_ hesitant of touching the submission system since that's probably our most critical thing. Maybe add a TODO in here?